source/opt/trim_capabilities_pass.cpp

```diff
@@ -0,0 +1,197 @@
+    const IRContext* context, const Instruction* instruction)>;
+
+// ============== Begin opcode handler implementations. =======================
+//
+// Adding support for a new capability should only require adding a new handler,
+// and updating the
+// kSupportedCapabilities/kUntouchableCapabilities/kFordiddenCapabilities lists.
+//
+// Handlers names follow the following convention:
+//  Handler_<Opcode>_<Capability>()
+
+static std::optional<spv::Capability> Handler_OpReadClockKHR_ShaderClockKHR(
+    const IRContext*, const Instruction*) {
+  // The simple presence of this opcode is enough to determine the capability
+  // requirement.
+  return spv::Capability::ShaderClockKHR;
+}
+
+// Opcode of interest to determine capabilities requirements.
+static const std::unordered_multimap<spv::Op, OpcodeHandler, UnsignedHash>
+    kOpcodeHandlers{
+        {spv::Op::OpReadClockKHR, Handler_OpReadClockKHR_ShaderClockKHR},
+    };
+
+// ==============  End opcode handler implementations.  =======================
+
+CapabilitySet TrimCapabilitiesPass::DetermineRequiredCapabilities() {
+  CapabilitySet required_capabilities;
+  const IRContext* ctx = context();
+
+  get_module()->ForEachInst(
+      [ctx, &required_capabilities](Instruction* instruction) {
+        auto[begin, end] = kOpcodeHandlers.equal_range(instruction->opcode());
+        std::for_each(begin, end,
+                      [&required_capabilities, ctx, instruction](auto& item) {
+                        auto result = item.second(ctx, instruction);
+                        if (result.has_value()) {
+                          required_capabilities.insert(*result);
+                        }
+                      });
+      });
+
+  return required_capabilities;
+}
+
+// Utility function to, given a list of capabilities, retrieve the associated
+// extensions.
+template <typename Container>
+static ExtensionSet GetExtensionsForCapabilities(
+    const Container& capabilities) {
+  ExtensionSet output;
+  for (auto capability : capabilities) {
+    auto extension = kCapabilityExtensionMapping.find(capability);
+    if (extension != kCapabilityExtensionMapping.end()) {
+      output.insert(extension->second);
+    }
+  }
+  return output;
+}
+
+void TrimCapabilitiesPass::TrimExtensionsForCapabilities(
+    const CapabilitySet& required_capabilities,
+    const std::vector<spv::Capability>& trimmed_capabilities) {
+  const ExtensionSet required_extensions =
+      GetExtensionsForCapabilities(required_capabilities);
+  const ExtensionSet trimmed_extensions =
+      GetExtensionsForCapabilities(trimmed_capabilities);
+
+  std::vector<Instruction*> instructions_to_kill;
+  for (auto& instruction : context()->extensions()) {
+    assert(instruction.NumOperands() == 1);
+    auto extension = instruction.GetOperand(0).AsString();
+
+    if (required_extensions.count(extension) != 0) {
+      continue;
+    }
+
+    if (trimmed_extensions.count(extension) != 0) {
```

Comment:
Add a comment explaining why you do not kill an extension even if it is not required by `required_capabilities`.